File: modules/debugger/src/main/java/com/falsepattern/zigbrains/debugger/toolchain/DebuggerAvailability.java
package com.falsepattern.zigbrains.debugger.toolchain;

import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.nio.file.Path;

public sealed interface DebuggerAvailability {
    enum Kind {
        Unavailable,
        NeedToDownload,
        NeedToUpdate,
        Bundled,
        Binaries
    }

    Kind kind();
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    final class Unavailable implements DebuggerAvailability {
        @Override
        public Kind kind() {
            return Kind.Unavailable;
        }
    }
    Unavailable Unavailable = new Unavailable();

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    final class NeedToDownload implements DebuggerAvailability {
        @Override
        public Kind kind() {
            return Kind.NeedToDownload;
        }
    }
    NeedToDownload NeedToDownload = new NeedToDownload();

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    final class NeedToUpdate implements DebuggerAvailability {
        @Override
        public Kind kind() {
            return Kind.NeedToUpdate;
        }
    }
    NeedToUpdate NeedToUpdate = new NeedToUpdate();

    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    final class Bundled implements DebuggerAvailability {
        @Override
        public Kind kind() {
            return Kind.Bundled;
        }
    }
    Bundled Bundled = new Bundled();

    sealed interface Binaries extends DebuggerAvailability {
        @Override
        default Kind kind() {
            return Kind.Binaries;
        }

        DebuggerKind binariesKind();
    }

    record LLDBBinaries(Path frameworkFile, Path frontendFile) implements Binaries {
        @Override
        public DebuggerKind binariesKind() {
            return DebuggerKind.LLDB;
        }
    }
    record GDBBinaries(Path gdbFile) implements Binaries {
        @Override
        public DebuggerKind binariesKind() {
            return DebuggerKind.GDB;
        }
    }
    record MSVCBinaries(Path msvcFile) implements Binaries {
        @Override
        public DebuggerKind binariesKind() {
            return DebuggerKind.MSVC;
        }
    }
}

File: cidr/src/main/kotlin/com/falsepattern/zigbrains/debugger/toolchain/DebuggerAvailability.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2024 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.debugger.toolchain

import java.nio.file.Path

sealed class DebuggerAvailability<out T> {
    data object Unavailable: DebuggerAvailability<Nothing>()
    data object NeedToDownload: DebuggerAvailability<Nothing>()
    data object NeedToUpdate: DebuggerAvailability<Nothing>()
    data object Bundled: DebuggerAvailability<Nothing>()
    data class Binaries<T> (val binaries: T): DebuggerAvailability<T>()
}

data class LLDBBinaries(val frameworkFile: Path, val frontendFile: Path)
data class GDBBinaries(val gdbFile: Path)
data class MSVCBinaries(val msvcFile: Path)